theano/scalar/basic.py

```diff
@@ -1928,22 +1928,24 @@ def c_code(self, node, name, inputs, outputs, sub):
 
         t = node.inputs[0].type.upcast(*[i.type for i in node.inputs[1:]])
         if t in imap(str, discrete_types):
+            x_div_y_pp = '(%(x)s / %(y)s)' % locals()
+            x_div_y_mp = '((-%(x)s) / %(y)s)' % locals()
+            x_mod_y_mp = 'THEANO_MACRO_MOD((-%(x)s), %(y)s)' % locals()
+            x_div_y_pm = '(%(x)s / (-%(y)s))' % locals()
+            x_mod_y_pm = 'THEANO_MACRO_MOD(%(x)s, (-%(y)s))' % locals()
+            x_div_y_mm = '((-%(x)s) / (-%(y)s))' % locals()
             # If we are in a gpuarray kernel, %(fail)s exits the kernel,
             # and we do not have any error report, and we cannot set
             # Python error messages either, so for now we just call the
             # cuda function, which return a binary pattern of all 1s.
-            check = dedent('''
-                #ifndef KERNEL
+            div_zero = dedent('''
+                #ifdef KERNEL
+                    %(z)s = %(x_div_y_pp)s;
+                #else
                     PyErr_SetString(PyExc_ZeroDivisionError, "integer division by zero");
                     %(fail)s
```

Comment:
Did you test that this remove the openmp bug? This don't change the code generated with openmp, the goto is still in the middle of the loop